clippy_lints/src/module_style.rs

```diff
@@ -0,0 +1,172 @@
+use std::{
+    ffi::OsString,
+    path::{Component, Path},
+};
+
+use rustc_ast::ast;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_lint::{EarlyContext, EarlyLintPass, LintContext};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{FileName, RealFileName, SourceFile, Span, SyntaxContext};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks that module layout uses only self named module files, bans mod.rs files.
+    ///
+    /// ### Why is this bad?
+    /// Having multiple module layout styles in a project can be confusing.
+    ///
+    /// ### Example
+    /// ```text
+    /// src/
+    ///   stuff/
+    ///     stuff_files.rs
+    ///     mod.rs
+    ///   lib.rs
+    /// ```
+    /// Use instead:
+    /// ```text
+    /// src/
+    ///   stuff/
+    ///     stuff_files.rs
+    ///   stuff.rs
+    ///   lib.rs
+    /// ```
+    pub MOD_FILES,
```

Comment:
thoughts on improving the lint name?

Reply:
`mod_module_files` and the other could be `self_named_module_files` :shrug: I was hoping when I looked up the RFC that introduced the new module system there would be a good name for it but there is not.

Reply:
@camsteffen thoughts?

Reply:
Hmmm. I think @DevinR528's idea is good. Another approach is `mod_file_within_dir` and `mod_file_outside_dir` but I don't like that better.